common/src/primitives/merkle/tree.rs

```diff
@@ -0,0 +1,560 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::BTreeSet, num::NonZeroUsize};
+
+use crypto::hash::StreamHasher;
+use itertools::Itertools;
+
+use crate::primitives::{
+    id::{default_hash, DefaultHashAlgoStream},
+    H256,
+};
+
+use super::{MerkleTreeAccessError, MerkleTreeFormError, MerkleTreeProofExtractionError};
+
+pub enum AdjacentLeavesIndices {
+    Alone(usize),
+    Together(usize, usize),
+}
+
+/// Merkle tree in the form of a vector, where the bottom leaves are the based, and the root is
+/// the last element.
+#[derive(Debug)]
+pub struct MerkleTree {
+    tree: Vec<H256>,
+}
+
+fn next_pow2(n: usize) -> usize {
+    if n == 0 {
+        return 1;
+    }
+    let leading_zeros = (n - 1).leading_zeros() as usize;
+    let active_bits = usize::BITS as usize - leading_zeros;
+    (1 << active_bits) as usize
+}
+
+fn create_merkletree_padding(elements: &[H256]) -> Vec<H256> {
```

Comment:
For this particular solution, yes, we can play all the games we want. But I want to point out that this is a security feature and solutions must be constrained and replaceable. For example, your last solution is a bad idea because if we ever deciede to pad with something else this format won't work. Pros and cons to security applications. I'll give the first solution with the iterators a shot, but I must constrain this to be deterministic and 100% testable and contained more than it being nice.